### PR TITLE
Add function to recover staked funds in misaligned motions

### PR DESCRIPTION
**Pull request by area**
Based on #1081, so no merging. Broadly speaking:

* Have this function declared in a separate file, with the (wrong) storage layout from v4 so that we can use normal-looking solidity to access the misaligned funds.
* Had to introduce IVotingReputation as we're spreading the functions across multiple files
* Slight alteration to implementation in the recovery functions, but hopefully not results. Basically, a motion must be finalized in order for people to be rewarded/punished based on which side they staked on, but because these motions no longer can advance state, I needed to switch to keying off of `finalized` boolean.

### CODE REVIEW

File: contracts/extensions/VotingReputation/VotingReputation.sol [hunk 1]
```diff
@@ -18,28 +18,17 @@
...
+import "./../../common/BasicMetaTransaction.sol";
+import "./../../common/ERC20Extended.sol";
+import "./../../patriciaTree/PatriciaTreeProofs.sol";
+import "./../../tokenLocking/ITokenLocking.sol";
+import "./../ColonyExtension.sol";
+import "./VotingReputationDataTypes.sol";
+
+
+contract VotingReputation is ColonyExtension, PatriciaTreeProofs, BasicMetaTransaction, VotingReputationDataTypes {
   uint256 constant UINT128_MAX = 2**128 - 1;
```
Comment (kronosapiens) [inline contracts/extensions/VotingReputation/VotingReputation.sol:32]: I would put the constants in `DataTypes` also

File: contracts/extensions/VotingReputation/VotingReputationMisalignedRecovery.sol [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+import "./../../colony/ColonyRoles.sol";
+import "./../../common/BasicMetaTransaction.sol";
+import "./../../common/ERC20Extended.sol";
+import "./../../patriciaTree/PatriciaTreeProofs.sol";
+import "./../../tokenLocking/ITokenLocking.sol";
+import "./../ColonyExtension.sol";
+import "./../../../lib/dappsys/math.sol";
+import "./VotingReputationDataTypes.sol";
+
+contract VotingReputationMisalignedRecovery is PatriciaTreeProofs, DSMath, DSAuth, VotingReputationDataTypes {
```
Comment (kronosapiens) [inline contracts/extensions/VotingReputation/VotingReputationMisalignedRecovery.sol:31]: Don't know if PatriciaTreeProofs is needed here

File: contracts/extensions/VotingReputation/VotingReputationMisalignedRecovery.sol [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  IColony colony;
+  bool deprecated;
+
+  ExtensionState state;
+
+  IColonyNetwork colonyNetwork;
+  ITokenLocking tokenLocking;
+  address token;
+
+  // All `Fraction` variables are stored as WADs i.e. fixed-point numbers with 18 digits after the radix. So
```
Comment (kronosapiens) [inline contracts/extensions/VotingReputation/VotingReputationMisalignedRecovery.sol:53]: I would delete all the variable descriptive comments in this file, it gives the impression that this file is to be "taken seriously"

File: contracts/extensions/VotingReputation/VotingReputationMisalignedRecovery.sol [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  You should have received a copy of the GNU General Public License
+  along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+pragma solidity 0.7.3;
+pragma experimental ABIEncoderV2;
+
+import "./../../colonyNetwork/IColonyNetwork.sol";
+import "./../../colony/ColonyRoles.sol";
+import "./../../common/BasicMetaTransaction.sol";
```
Comment (kronosapiens) [inline contracts/extensions/VotingReputation/VotingReputationMisalignedRecovery.sol:23]: Are we using this?

File: contracts/extensions/VotingReputation/VotingReputationMisalignedRecovery.sol [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+  // Setting this to anything other than 1 will mean it is likely not all those eligible to vote will be able to do so.
+
+  // All `Period` variables are second-denominated
+
+  uint256 stakePeriod; // Length of time for staking
+  uint256 submitPeriod; // Length of time for submitting votes
+  uint256 revealPeriod; // Length of time for revealing votes
+  uint256 escalationPeriod; // Length of time for escalating after a vote
+
+  // Here we deliberately recreate the misalignment in the storage slots, so solidity can correctly
```
Comment (kronosapiens) [inline contracts/extensions/VotingReputation/VotingReputationMisalignedRecovery.sol:74]: Except the comments to do with the misaligned storage

**Review by kronosapiens — COMMENTED** [review]
I would rename the directory `votingReputation` to be consistent with other parts of the codebase.

File: contracts/extensions/VotingReputation/VotingReputationMisalignedRecovery.sol (hunk 5 above)
Comment (kronosapiens) [reply contracts/extensions/VotingReputation/VotingReputationMisalignedRecovery.sol:74]: I would be inclined also to re-name the unused variables (unused in the two functions we implement) with dummy names to further emphasize that they are not used for anything.

File: contracts/extensions/VotingReputation/IVotingReputation.sol [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+  /// @return The voter reward
+  function getVoterRewardRange(uint256 _motionId, uint256 _voterRep, address _voterAddress) external view returns (uint256, uint256) ;
+  /// @notice Get the staker reward
+  /// @param _motionId The id of the motion
+  /// @param _staker The staker's address
+  /// @param _vote The vote (0 = NAY, 1 = YAY)
+  /// @return The staker reward and the reputation penalty (if any)
+  function getStakerReward(uint256 _motionId, address _staker, uint256 _vote) external view returns (uint256, uint256);
+
+  function createClaimDelayAction(bytes memory action, uint256 value)
```
[the remark this answers is not in the document]
Comment (area) [reply contracts/extensions/VotingReputation/IVotingReputation.sol:309]: It is, and has been for a while. Not intending to do a pass of that type of thing here, but we can pick it up down the line.

File: contracts/extensions/VotingReputation/VotingReputation.sol (hunk 1 above)
Comment (area) [reply contracts/extensions/VotingReputation/VotingReputation.sol:32]: out

**Comment by area** [comment]
This was included in #1084, but if you want more changes will need rebasing and more work from me before being merged in to develop.

**Pull request closed without merging**